### DISCUSSION TITLE
Allow virtual methods to be annotated as "must be called"

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/csharplang#6432] [Language Ideas]
This question comes up from time to time, we override some base method and implement some logic then ask ourselves "Do we need to call the base method at all?".

A recent example is `SetParametersAsync` in the base class `ComponentBase` used in `Blazor`. If the base method is _not_ called in the overridden method then the component lifecycle is no longer honored.

In cases where the designer requires the base method to be called there is no way to communicate that today other than through documentation.

Is there an merit therefore in _marking_ a base method in some way, so that it can be determined (either by the compiler or perhaps by an analyzer) that the base method must be called?

**Comment by CyrusNajmabadi** [comment]
> In cases where the designer requires the base method to be called there is no way to communicate that today other than through documentation.

This can be communicated through an analyzer as an example.

> Is there an merit therefore in marking a base method in some way, so that it can be determined (either by the compiler or perhaps by an analyzer) that the base method must be called?

Yup, i could see someone coming up with an attribute that an analyzer then checked for.  Seems entirely reasonable.  This would not require a lang change though :)

**Comment by Joe4evr** [comment]
As an alternative to an analyzer, you could design your code with the Template Method Pattern. (Obviously, this is out of your control when it comes to code you don't own, but it's good to take into account with your own code.)

Upsides:
* Overrides don't need to worry about calling the base implementation (and *when* they do so), because required code can always run when the base type needs to
* Personally, I find that *public* virtual methods should be considered somewhat of an anti-pattern, since
  * *Inheriting* classes should be the primary cases caring that a member is virtual
  * In C# specifically, the `?.` operator actually makes changing a non-virtual member to virtual between releases a breaking change (the compiler emits `call` rather than `callvirt` in the first case; consuming code would need to recompile to get the correct behavior)

Downsides:
* (Usually) Requires coming up with different names between the public and protected members (though prefixing with `On` is common enough)
